src/EFCore.PG/Query/ExpressionVisitors/NpgsqlSqlTranslatingExpressionVisitor.cs

```diff
@@ -122,101 +81,80 @@ protected override Expression VisitBinary(BinaryExpression expression)
         /// An '= ANY' expression or null.
         /// </returns>
         [CanBeNull]
-        protected virtual Expression VisitEqualsAny([NotNull] SubQueryExpression expression)
-        {
-            var subQueryModel = expression.QueryModel;
-            var fromExpression = subQueryModel.MainFromClause.FromExpression;
+        protected override Expression VisitSubQuery(SubQueryExpression expression)
+            => base.VisitSubQuery(expression) ?? VisitArraySubQuery(expression);
 
-            var properties = MemberAccessBindingExpressionVisitor.GetPropertyPath(
-                fromExpression, _queryModelVisitor.QueryCompilationContext, out _);
+        /// <inheritdoc />
+        protected override Expression VisitUnary(UnaryExpression expression)
+            => Visit(expression.Operand) is Expression operand
+                ? Expression.MakeUnary(expression.NodeType, operand, expression.Type)
+                : base.VisitUnary(expression);
 
-            if (properties.Count == 0)
-                return null;
-            var lastPropertyType = properties[properties.Count - 1].ClrType;
-            if (lastPropertyType.IsArray && lastPropertyType.GetArrayRank() == 1 && subQueryModel.ResultOperators.Count > 0)
-            {
-                // Translate someArray.Length
-                if (subQueryModel.ResultOperators.First() is CountResultOperator)
```

Comment:
I'm a bit surprised array indexing works without the special handling that you removed - are we sure the tests are exercising it correctly etc.?

Note that there's a distinction between the translation phase (represented by `NpgsqlSqlTranslatingExpressionVisitor`) and the SQL generation phase. I admit I'm not totally totally clear on the details of this separation, but it seems more correct to have a standard `ArrayLength` node in the expression tree as before. This way, when our SQL generator starts its job, it has an input expression tree that is as close as possible to what a real C# expression tree would look like. We need to review all this and understand it better.